client/cli/src/commands/runcmd.rs

```diff
@@ -428,13 +414,12 @@ impl CliConfiguration for RunCmd {
 		)?))
 	}
 
-	fn offchain_worker(&self, role: &Role) -> Result<bool> {
-		Ok(match (&self.offchain_worker, role) {
-			(OffchainWorkerEnabled::WhenValidating, Role::Authority { .. }) => true,
-			(OffchainWorkerEnabled::Always, _) => true,
-			(OffchainWorkerEnabled::Never, _) => false,
-			(OffchainWorkerEnabled::WhenValidating, _) => false,
-		})
+	fn offchain_worker_config(&self) -> OffchainWorkerConfig {
+		let is_dev = self.is_dev();
+		let role = self.role(is_dev);
```

Comment:
is_dev and role must be passed by arguments. You can change the signature of offchain_worker_config to provide them (in config.rs) and change create_configuration (in config.rs) so they are given in arguments.

Reply:
Example here with pruning: https://github.com/paritytech/substrate/blob/master/client/cli/src/config.rs#L400

You can totally inspire on pruning because it is the exact same use case than yours:
- it has a *Params structopt
- it defines a _params method in CliConfiguration to get this params
- the method pruning uses that _params method to provide the default

I hope this will help.